api/src/wfl/module/covid.clj

```diff
@@ -152,6 +152,24 @@
                              {:workload workload})))
     (if-not (or stopped finished) (stop! workload (utc-now)) workload)))
 
+(defn ^:private retry-covid-workload
+  "Retry/resubmit the `workflows` managed by the `workload` and return the
+   workload that manages the new workflows."
+  [{:keys [started id executor] :as workload} workflows]
+  (when-not started
+    (throw (UserException. "Cannot retry workload before it's been started."
+                           {:workload workload})))
+  (when (empty? workflows)
+    (throw (UserException. "No workflows specified to retry in workload."
+                           {:workload workload})))
```

Comment:
If you're going to throw, throw sooner upstream - that way you can make the error message more meaningful, i.e.
```clojure
(throw (UserException. "Retry failed: no workflows found matching status" {:status status}))
```
(this will cause another issue though in that the `UserException` handler will try to convert this `:status` into an HTTP status code, so need to change that to test if it has a numeric `:status`, or change it to test for something else like `:http-status` or `:code` or `:status-code` or something)

Reply:
Also, the fact that a workload implementation doesn't support retries should be secondary - you've already implied this by checking for `Succeeded`

Reply:
(To make sure I understand) "sooner upstream" should be within `handlers/post-retry`.

That's a good point that by checking if the status is `Succeeded` I've already shown that checking whether retries are implemented isn't a check that needs to happen first.  I agree that a more meaningful error message outweighs the changes to system tests.

(I did get burned by the `:status` naming collision earlier, thanks for flagging it.)

Reply:
Yeah exactly - in `post-retry` :)